### ISSUE TITLE
[receiver/prometheusremotewrite] Flaky test

### ISSUE THREAD

**Issue by ArthurSens** [open-telemetry/opentelemetry-collector-contrib#36654]
### Component(s)

receiver/prometheusremotewrite

### Describe the issue you're reporting

I've just noticed we have some flaky tests in main: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/12146357038/job/33870088465

```
=== FAIL: . TestComponentLifecycle/metrics-lifecycle (0.00s)
    generated_component_test.go:65: 
        	Error Trace:	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/prometheusremotewritereceiver/generated_component_test.go:65
        	Error:      	Received unexpected error:
        	            	failed to create prometheus remote-write listener: listen tcp 127.0.0.1:9090: bind: address already in use
        	Test:       	TestComponentLifecycle/metrics-lifecycle

=== FAIL: . TestComponentLifecycle (0.00s)

=== FAIL: . TestHandlePRWContentTypeNegotiation/no_content_type (0.00s)
    receiver_test.go:90: 
        	Error Trace:	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/prometheusremotewritereceiver/receiver_test.go:90
        	Error:      	Received unexpected error:
        	            	Post "http://localhost:9090/api/v1/write": dial tcp [::1]:9090: connect: connection refused
        	Test:       	TestHandlePRWContentTypeNegotiation/no_content_type

=== FAIL: . TestHandlePRWContentTypeNegotiation (0.00s)
    receiver_test.go:[43](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/12146357038/job/33870088465#step:7:44): 
        	Error Trace:	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/prometheusremotewritereceiver/receiver_test.go:36
        	            				/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/prometheusremotewritereceiver/receiver_test.go:43
        	Error:      	Received unexpected error:
        	            	failed to create prometheus remote-write listener: listen tcp 127.0.0.1:9090: bind: address already in use
        	Test:       	TestHandlePRWContentTypeNegotiation
```

**Comment by ArthurSens** [comment]
Yeah, that's probably enough :)

**Comment by pjanotti** [comment]
Another hit https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/12815609779/job/35734659123#step:8:37

**Comment by pjanotti** [comment]
> maybe we need to change tests to use a random port?

Very wild guess without looking at the code, but, I saw this some symptom many times on Windows tests. One possibility is that the component has a goroutine that holds the port/server and while the goroutine is terminated when the component is shut down it may do so asynchronously so the next test may find the port is in use. Once more I'm shooting from the hip here ...

**Comment by pjanotti** [comment]
another hit https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/12836136973/job/35797126187

**Comment by ArthurSens** [comment]
We're hardcoding the port during tests:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0abff4c84b662c9f63187a65d2097fe6787f91cf/receiver/prometheusremotewritereceiver/receiver_test.go#L116

We probably want to use `httptest.ServeHTTP` instead, so we pick a random port. @perebaj, is this another you'd like to fix? Again no pressure, my family issues should be over next week. I can tackle this if you can't :)

**Comment by perebaj** [comment]
On my list. Don't worry. ✌🏽

**Comment by pjanotti** [comment]
@perebaj @ArthurSens another thing to consider: sometimes the type of failure is because a prior test is still holding the port when another test is trying to open it again. Perhaps this was already helped by #37430 - I'm not sure at this point.

**Comment by perebaj** [comment]
> [@perebaj](https://github.com/perebaj) [@ArthurSens](https://github.com/ArthurSens) another thing to consider: sometimes the type of failure is because a prior test is still holding the port when another test is trying to open it again. Perhaps this was already helped by [#37430](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37430) - I'm not sure at this point.

Thanks for the tip, I will consider it

**Comment by adwait-godbole** [comment]
Hi @perebaj, I hope you're doing well! I wanted to ask if you’ve already started working on fixing up this issue? I'm an LFX enthusiast, and since `PrometheusRemoteWrite` receiver is listed as a project for this term, I’ve been trying to get familiar with the code so far. If you’ve already started work on it, I wouldn’t mind stepping aside, but if not, would you be okay with me picking it up? Also, if you have any inputs or suggestions, I’d greatly appreciate them. Thanks!

**Comment by perebaj** [comment]
> Hi [@perebaj](https://github.com/perebaj), I hope you're doing well! I wanted to ask if you’ve already started working on fixing up this issue? I'm an LFX enthusiast, and since `PrometheusRemoteWrite` receiver is listed as a project for this term, I’ve been trying to get familiar with the code so far. If you’ve already started work on it, I wouldn’t mind stepping aside, but if not, would you be okay with me picking it up? Also, if you have any inputs or suggestions, I’d greatly appreciate them. Thanks!

Hey ✌🏽. I'm already started to work on it.

Maybe this issue would be nice for you https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33661

I'm working on `addTypeSum. But there are a lot of open TODOs to address.

**Comment by adwait-godbole** [comment]
Thanks @perebaj for pointing me over there at #33661.

> I'm working on `addTypeSum. But there are a lot of open TODOs to address.

 Shall I work on top of this already open PR of yours at #37156? 

Also, when using `PrometheusRemoteWriteExporter`, which receiver should I use to test on my local machine? 😅

**Comment by perebaj** [comment]
> Shall I work on top of this already open PR of yours at [#37156](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37156)?

I don't think so, I'm finishing it, just missing a test case. I think that [this one](https://github.com/perebaj/opentelemetry-collector-contrib/blob/fac8d2d57ef891b4fee799652f75b1f9f06decb3/pkg/translator/prometheusremotewrite/metrics_to_prw_v2.go#L92) would be a good idea.

The approach that I'm following is using the [previous implementation as reference](https://github.com/perebaj/opentelemetry-collector-contrib/blob/5af8c701ef426e08b30cb976933fe4256ce51233/pkg/translator/prometheusremotewrite/metrics_to_prw.go#L97), but using prw_v2 functions. WDYT?
